llvm-external-projects/iree-dialects/python/mlir/dialects/iree_pydm/rtl/modules/macros.py

```diff
@@ -0,0 +1,121 @@
+from ..... import ir
+
+
+@def_ir_macro_intrinsic
+def get_type_code(stage: ImportStage, value: ir.Value) -> ir.Value:
+  """Gets the TypeCode (see C++ BuiltinTypeCode) associated with a value."""
+  return d.GetTypeCodeOp(d.IntegerType.get(), value).result
+
+
+@def_ir_macro_intrinsic
+def is_type(stage: ImportStage, value: ir.Value,
+            type_code: ir.Value) -> ir.Value:
+  """Efficiently checks whether a value has a given type code."""
+  ic = stage.ic
+  if not d.IntegerType.isinstance(type_code.type):
+    ic.abort(f"is_type() macro must be called with a constant type_code. "
+             f"Got {type_code}")
+  actual_type_code = get_type_code(stage, value)
+  cmp_result = d.ApplyCompareOp(d.BoolType.get(), ir.StringAttr.get("eq"),
+                                type_code, actual_type_code).result
+  return cmp_result
+
+
+@def_ir_macro_intrinsic
+def get_numeric_promotion_order(stage: ImportStage,
+                                value: ir.Value) -> ir.Value:
+  """Gets the numeric promotion order.
+
+  See get_numeric_promotion_order op.
+  """
+  return d.GetNumericPromotionOrderOp(d.IntegerType.get(), value).result
+
+
+@def_ir_macro_intrinsic
+def promote_numeric_to_integer(stage: ImportStage, value: ir.Value) -> ir.Value:
+  """Promotes the value to IntegerType."""
+  return d.PromoteNumericOp(d.IntegerType.get(), value).result
+
+
+@def_ir_macro_intrinsic
+def promote_numeric_to_real(stage: ImportStage, value: ir.Value) -> ir.Value:
+  """Promotes the value to RealType."""
+  return d.PromoteNumericOp(d.RealType.get(), value).result
+
+
+@def_ir_macro_intrinsic
+def unbox_unchecked_bool(stage: ImportStage, value: ir.Value) -> ir.Value:
+  """Unboxes an object value to a bool, not checking for success."""
+  return d.UnboxOp(d.ExceptionResultType.get(), d.BoolType.get(),
+                   value).primitive
+
+
+@def_ir_macro_intrinsic
+def unbox_unchecked_integer(stage: ImportStage, value: ir.Value) -> ir.Value:
+  """Unboxes an object value to an integer, not checking for success."""
+  return d.UnboxOp(d.ExceptionResultType.get(), d.IntegerType.get(),
+                   value).primitive
+
+
+@def_ir_macro_intrinsic
+def unbox_unchecked_real(stage: ImportStage, value: ir.Value) -> ir.Value:
+  """Unboxes an object value to a real, not checking for success."""
+  return d.UnboxOp(d.ExceptionResultType.get(), d.RealType.get(),
+                   value).primitive
+
+
+@def_ir_macro_intrinsic
+def raw_compare_eq(stage: ImportStage, left: ir.Value,
+                   right: ir.Value) -> ir.Value:
+  """Emits an ApplyCompareOp for 'eq'."""
+  return d.ApplyCompareOp(d.BoolType.get(), ir.StringAttr.get("eq"), left,
+                          right).result
+
+
+@def_ir_macro_intrinsic
+def raw_compare_gt(stage: ImportStage, left: ir.Value,
+                   right: ir.Value) -> ir.Value:
+  """Emits an ApplyCompareOp for 'gt'."""
+  return d.ApplyCompareOp(d.BoolType.get(), ir.StringAttr.get("gt"), left,
```

Comment:
this is pretty nice syntax - ooc does the python intellisense in vscode complete this all?

Reply:
Partially. It does for generated code but could use some more typing to be better. I still need to generate .pyi files for the native stuff but keep hoping someone else does it (versus just complaining).